bindings/java/src/org/sleuthkit/datamodel/BlackboardArtifact.java

```diff
@@ -384,6 +385,41 @@ public void addAttributes(Collection<BlackboardAttribute> attributes) throws Tsk
 		attrsCache.addAll(attributes);
 	}
 
+	/**
+	 * Adds a collection of attributes to this artifact in a single operation
+	 * (faster than adding each attribute individually). This function also
+	 * executes in the inserts within the transaction context.
+	 *
+	 * @param attributes		      The collection of attributes.
+	 * @param caseDbTransaction the transaction in the scope of which the
+	 *                          operation is to be performed, managed by the
+	 *                          caller
+	 *
+	 * @throws TskCoreException If an error occurs and the attributes were not
+	 *                          added to the artifact.
+	 */
+	public void addAttributes(Collection<BlackboardAttribute> attributes, final SleuthkitCase.CaseDbTransaction caseDbTransaction) throws TskCoreException {
+
+		if (attributes.isEmpty()) {
+			return;
+		}
+		if (caseDbTransaction == null) {
+			addAttributes(attributes);
+			return;
+		}
+		try {
+			for (final BlackboardAttribute attribute : attributes) {
+				attribute.setArtifactId(artifactId);
+				attribute.setCaseDatabase(getSleuthkitCase());
+				getSleuthkitCase().addBlackBoardAttribute(attribute, artifactTypeId, caseDbTransaction.getConnection());
+			}
+			attrsCache.addAll(attributes);
+		} catch (SQLException ex) {
+			throw new TskCoreException("Error adding blackboard attributes", ex);
```

Comment:
To help us if we ever have to do debugging that involves this code, please add the object ID and/or artifact ID of the artifact to the error message.